# ONLINE RETAIL BIG DATA PROJECT USING PYSPARK

Data description found in [this link](https://archive.ics.uci.edu/ml/datasets/Online%20Retail).

# Data Processing

#### Load Data

In [0]:
# File location and type
file_location = "/FileStore/tables/OnlineRetail.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

InvoiceNo StockCode Description Quantity InvoiceDate UnitPrice CustomerID Country 536365 85123A WHITE HANGING HEART T-LIGHT HOLDER 6 01/12/2010 08:26 2.55 17850 United Kingdom 536365 71053 WHITE METAL LANTERN 6 01/12/2010 08:26 3.39 17850 United Kingdom 536365 84406B CREAM CUPID HEARTS COAT HANGER 8 01/12/2010 08:26 2.75 17850 United Kingdom 536365 84029G KNITTED UNION FLAG HOT WATER BOTTLE 6 01/12/2010 08:26 3.39 17850 United Kingdom 536365 84029E RED WOOLLY HOTTIE WHITE HEART. 6 01/12/2010 08:26 3.39 17850 United Kingdom 536365 22752 SET 7 BABUSHKA NESTING BOXES 2 01/12/2010 08:26 7.65 17850 United Kingdom 536365 21730 GLASS STAR FROSTED T-LIGHT HOLDER 6 01/12/2010 08:26 4.25 17850 United Kingdom 536366 22633 HAND WARMER UNION JACK 6 01/12/2010 08:28 1.85 17850 United Kingdom 536366 22632 HAND WARMER RED POLKA DOT 6 01/12/2010 08:28 1.85 17850 United Kingdom 536367 84879 ASSORTED COLOUR BIRD ORNAMENT 32 01/12/2010 08:34 1.69 13047 United Kingdom 536367 22745 POPPY'S PLAYHOUSE BEDROOM 6 01/12/2010 08:34 2.1 13047 United Kingdom 536367 22748 POPPY'S PLAYHOUSE KITCHEN 6 01/12/2010 08:34 2.1 13047 United Kingdom 536367 22749 FELTCRAFT PRINCESS CHARLOTTE DOLL 8 01/12/2010 08:34 3.75 13047 United Kingdom 536367 22310 IVORY KNITTED MUG COSY 6 01/12/2010 08:34 1.65 13047 United Kingdom 536367 84969 BOX OF 6 ASSORTED COLOUR TEASPOONS 6 01/12/2010 08:34 4.25 13047 United Kingdom 536367 22623 BOX OF VINTAGE JIGSAW BLOCKS 3 01/12/2010 08:34 4.95 13047 United Kingdom 536367 22622 BOX OF VINTAGE ALPHABET BLOCKS 2 01/12/2010 08:34 9.95 13047 United Kingdom 536367 21754 HOME BUILDING BLOCK WORD 3 01/12/2010 08:34 5.95 13047 United Kingdom 536367 21755 LOVE BUILDING BLOCK WORD 3 01/12/2010 08:34 5.95 13047 United Kingdom 536367 21777 RECIPE BOX WITH METAL HEART 4 01/12/2010 08:34 7.95 13047 United Kingdom 536367 48187 DOORMAT NEW ENGLAND 4 01/12/2010 08:34 7.95 13047 United Kingdom 536368 22960 JAM MAKING SET WITH JARS 6 01/12/2010 08:34 4.25 13047 United Kingdom 536368 22913 RED COAT RACK PARIS FASHION 3 01/12/2010 08:34 4.95 13047 United Kingdom 536368 22912 YELLOW COAT RACK PARIS FASHION 3 01/12/2010 08:34 4.95 13047 United Kingdom 536368 22914 BLUE COAT RACK PARIS FASHION 3 01/12/2010 08:34 4.95 13047 United Kingdom 536369 21756 BATH BUILDING BLOCK WORD 3 01/12/2010 08:35 5.95 13047 United Kingdom 536370 22728 ALARM CLOCK BAKELIKE PINK 24 01/12/2010 08:45 3.75 12583 France 536370 22727 ALARM CLOCK BAKELIKE RED 24 01/12/2010 08:45 3.75 12583 France 536370 22726 ALARM CLOCK BAKELIKE GREEN 12 01/12/2010 08:45 3.75 12583 France 536370 21724 PANDA AND BUNNIES STICKER SHEET 12 01/12/2010 08:45 0.85 12583 France 536370 21883 STARS GIFT TAPE 24 01/12/2010 08:45 0.65 12583 France 536370 10002 INFLATABLE POLITICAL GLOBE 48 01/12/2010 08:45 0.85 12583 France 536370 21791 VINTAGE HEADS AND TAILS CARD GAME 24 01/12/2010 08:45 1.25 12583 France 536370 21035 SET/2 RED RETROSPOT TEA TOWELS 18 01/12/2010 08:45 2.95 12583 France 536370 22326 ROUND SNACK BOXES SET OF4 WOODLAND 24 01/12/2010 08:45 2.95 12583 France 536370 22629 SPACEBOY LUNCH BOX 24 01/12/2010 08:45 1.95 12583 France 536370 22659 LUNCH BOX I LOVE LONDON 24 01/12/2010 08:45 1.95 12583 France 536370 22631 CIRCUS PARADE LUNCH BOX 24 01/12/2010 08:45 1.95 12583 France 536370 22661 CHARLOTTE BAG DOLLY GIRL DESIGN 20 01/12/2010 08:45 0.85 12583 France 536370 21731 RED TOADSTOOL LED NIGHT LIGHT 24 01/12/2010 08:45 1.65 12583 France 536370 22900 SET 2 TEA TOWELS I LOVE LONDON 24 01/12/2010 08:45 2.95 12583 France 536370 21913 VINTAGE SEASIDE JIGSAW PUZZLES 12 01/12/2010 08:45 3.75 12583 France 536370 22540 MINI JIGSAW CIRCUS PARADE 24 01/12/2010 08:45 0.42 12583 France 536370 22544 MINI JIGSAW SPACEBOY 24 01/12/2010 08:45 0.42 12583 France 536370 22492 MINI PAINT SET VINTAGE 36 01/12/2010 08:45 0.65 12583 France 536370 POST POSTAGE 3 01/12/2010 08:45 18.0 12583 France 536371 22086 PAPER CHAIN KIT 50'S CHRISTMAS 80 01/12/2010 09:00 2.55 13748 United Kingdom 536372 22632 HAND WARMER RED POLKA DOT 6 0

#### Missing values

In [0]:
# find missing values in all columns of the dataframe

from pyspark.sql.functions import col,isnan,when,count

df_missing = df.select([count(when(col(c).contains('None') | col(c).contains('NULL') | (col(c) == '' ) | col(c).isNull() | isnan(c), c)).alias(c) for c in df.columns])
df_missing.show()

+---------+---------+-----------+--------+-----------+---------+----------+-------+
|InvoiceNo|StockCode|Description|Quantity|InvoiceDate|UnitPrice|CustomerID|Country|
+---------+---------+-----------+--------+-----------+---------+----------+-------+
|        0|        0|       1454|       0|          0|        0|    135080|      0|
+---------+---------+-----------+--------+-----------+---------+----------+-------+



The column **CustomeID** contains **135,080** missing values. We shall drop all rows with missing values since we cannot attribute these transactions to customers - our focus in this project.

In [0]:
# drop rows with missing values in any column

df = df.na.drop()

#### Data type conversions

In [0]:
# convert InvoiceDate to a date format

from pyspark.sql.functions import to_date

df1 = df.select('InvoiceNo','StockCode','Description','Quantity','UnitPrice','CustomerID','Country', to_date(df.InvoiceDate, 'dd/MM/yyyy HH:mm').alias('dateFormatted'))

#### Duplicates

In [0]:
# let us keep only the distinct rows and drop any duplicates

df1 = df1.distinct()

# Customer Segmentation

#### Number of customers/products/countries

In [0]:
n_customers = df1.select("CustomerID").distinct().count()
n_products = df1.select("StockCode").distinct().count()
n_countries = df1.select("Country").distinct().count()

print("Number of customers: ", n_customers)
print("Number of products: ", n_products)
print("Number of countries: ", n_countries)

Number of customers:  4372
Number of products:  3684
Number of countries:  37


#### Products per transaction

In [0]:
from pyspark.sql.functions import count, col, desc

df1.groupBy('InvoiceNo').agg(count('StockCode').alias('no_products')).sort(desc('no_products')).show()

+---------+-----------+
|InvoiceNo|no_products|
+---------+-----------+
|   576339|        542|
|   579196|        533|
|   580727|        529|
|   578270|        442|
|   573576|        435|
|   567656|        421|
|   567183|        391|
|   575607|        377|
|   571441|        364|
|   570488|        353|
|   572552|        352|
|   568346|        335|
|   569246|        285|
|   547063|        281|
|   562031|        274|
|   570672|        259|
|   554098|        256|
|   543040|        244|
|   562046|        219|
|   569897|        218|
+---------+-----------+
only showing top 20 rows



#### Transactions per country

In [0]:
from pyspark.sql.functions import count, col, desc

transactions = df1.groupBy('Country').agg(count('InvoiceNo').alias('no_transactions')).sort(desc('no_transactions'))
display(transactions)

Country no_transactions United Kingdom 356727 Germany 9480 France 8475 EIRE 7475 Spain 2528 Netherlands 2371 Belgium 2069 Switzerland 1877 Portugal 1471 Australia 1258 Norway 1086 Italy 803 Channel Islands 757 Finland 695 Cyprus 611 Sweden 461 Austria 401 Denmark 389 Japan 358 Poland 341 USA 291 Israel 247 Unspecified 241 Singapore 229 Iceland 182 Canada 151 Greece 146 Malta 127 United Arab Emirates 68 European Community 61 RSA 58 Lebanon 45 Lithuania 35 Brazil 32 Czech Republic 30 Bahrain 17 Saudi Arabia 10

Output can only be rendered in Databricks

#### Time series of transactions per month

In [0]:
# transactions per month
from pyspark.sql.functions import month, mean, sum, asc, countDistinct

transactionsTS = df1.groupBy('dateFormatted').agg(countDistinct('InvoiceNo').alias('no_transactions')).sort(asc('dateFormatted'))
transactionsByMonth = transactionsTS.groupBy(month("dateFormatted").alias("month")).agg(sum("no_transactions").alias("transactions")).sort(asc('month'))
display(transactionsByMonth)

month transactions 1 1236 2 1202 3 1619 4 1384 5 1849 6 1707 7 1593 8 1544 9 2078 10 2263 11 3086 12 2629

Output can only be rendered in Databricks

#### Canceled orders
- The invoice numbers of transactions that were cancelled beging with 'C' as indicated in the description of the dataset

In [0]:
cancelled = df1.filter(col("InvoiceNo").contains("C"))
display(cancelled)

InvoiceNo StockCode Description Quantity UnitPrice CustomerID Country dateFormatted C536543 22355 CHARLOTTE BAG SUKI DESIGN -2 0.85 17841 United Kingdom 2010-12-01 C537320 35400 WOODEN BOX ADVENT CALENDAR -1 8.95 14867 United Kingdom 2010-12-06 C538713 22179 SET 10 LIGHTS NIGHT OWL -1 6.75 16570 United Kingdom 2010-12-14 C539956 35004C SET OF 3 COLOURED FLYING DUCKS -15 4.65 12980 United Kingdom 2010-12-23 C540006 21306 SET/4 DAISY MIRROR MAGNETS -1 2.1 14606 United Kingdom 2011-01-04 C540417 22197 SMALL POPCORN HOLDER -10 0.85 13680 United Kingdom 2011-01-07 C541009 22411 JUMBO SHOPPER VINTAGE RED PAISLEY -1 1.95 17581 United Kingdom 2011-01-13 C536812 22130 PARTY CONE CHRISTMAS DECORATION -144 0.72 16546 United Kingdom 2010-12-02 C537797 22501 PICNIC BASKET WICKER LARGE -1 8.5 13113 United Kingdom 2010-12-08 C538341 21647 ASSORTED TUTTI FRUTTI LARGE PURSE -1 2.1 15514 United Kingdom 2010-12-10 C539402 21527 RED RETROSPOT TRADITIONAL TEAPOT -1 6.95 17315 United Kingdom 2010-12-17 C539947 21363 HOME SMALL WOOD LETTERS -6 4.95 13854 United Kingdom 2010-12-23 C541416 22327 ROUND SNACK BOXES SET OF 4 SKULLS -1 2.95 14606 United Kingdom 2011-01-17 C536814 72799E IVORY PILLAR CANDLE SILVER FLOCK -6 2.95 13958 United Kingdom 2010-12-02 C537656 21258 VICTORIAN SEWING BOX LARGE -12 10.95 15100 United Kingdom 2010-12-07 C539945 22665 RECIPE BOX BLUE SKETCHBOOK DESIGN -3 2.95 17069 United Kingdom 2010-12-23 C540164 20914 SET/5 RED RETROSPOT LID GLASS BOWLS -6 2.95 14911 EIRE 2011-01-05 C540307 22212 FOUR HOOK WHITE LOVEBIRDS -12 2.1 15823 United Kingdom 2011-01-06 C540417 20712 JUMBO BAG WOODLAND ANIMALS -30 1.95 13680 United Kingdom 2011-01-07 C540837 71477 COLOUR GLASS. STAR T-LIGHT HOLDER -4 2.75 14733 United Kingdom 2011-01-11 C540942 37448 CERAMIC CAKE DESIGN SPOTTED MUG -1 1.49 13113 United Kingdom 2011-01-12 C541231 22059 CERAMIC STRAWBERRY DESIGN MUG -1 1.49 13455 United Kingdom 2011-01-14 C538122 22444 GROW YOUR OWN PLANT IN A CAN -1 1.25 14696 United Kingdom 2010-12-09 C536854 22158 3 HEARTS HANGING DECORATION RUSTIC -1 2.95 15240 United Kingdom 2010-12-03 C538693 79164 BLACK WINE GLASS -3 1.65 16558 United Kingdom 2010-12-14 C541499 22766 PHOTO FRAME CORNICE -1 2.95 13693 United Kingdom 2011-01-18 C538642 22636 CHILDS BREAKFAST SET CIRCUS PARADE -1 8.5 12476 Germany 2010-12-13 C539066 37449 CERAMIC CAKE STAND + HANGING CAKES -1 9.95 17912 United Kingdom 2010-12-15 C540109 85071A BLUE CHARLIE+LOLA PERSONAL DOORSIGN -150 2.95 17389 United Kingdom 2011-01-04 C536391 22553 PLASTERS IN TIN SKULLS -24 1.65 17548 United Kingdom 2010-12-01 C538059 22698 PINK REGENCY TEACUP AND SAUCER -6 2.95 14205 United Kingdom 2010-12-09 C539031 22202 MILK PAN PINK POLKADOT -1 3.75 12647 Germany 2010-12-15 C540378 21445 12 PINK ROSE PEG PLACE SETTINGS -4 1.25 13715 United Kingdom 2011-01-06 C540390 22701 PINK DOG BOWL -2 2.95 16191 United Kingdom 2011-01-06 C537132 22761 CHEST 7 DRAWER MA CAMPAGNE -2 21.95 15221 United Kingdom 2010-12-05 C539710 82483 WOOD 2 DRAWER CABINET WHITE FINISH -1 5.95 14769 United Kingdom 2010-12-21 C539719 82613C METAL SIGN,CUPCAKE SINGLE HOOK -20 0.42 14643 United Kingdom 2010-12-21 C540944 22822 CREAM WALL PLANTER HEART SHAPED -1 5.95 16554 United Kingdom 2011-01-12 C538085 21314 SMALL GLASS HEART TRINKET POT -1 2.1 16350 United Kingdom 2010-12-09 C540837 21733 RED HANGING HEART T-LIGHT HOLDER -1 2.55 14733 United Kingdom 2011-01-11 C538341 22728 ALARM CLOCK BAKELIKE PINK -1 3.75 15514 United Kingdom 2010-12-10 C538804 84945 MULTI COLOUR SILVER T-LIGHT HOLDER -2 0.85 13769 United Kingdom 2010-12-14 C538882 21892 TRADITIONAL WOODEN CATCH CUP GAME -12 1.25 12797 Portugal 2010-12-14 C540535 20914 SET/5 RED RETROSPOT LID GLASS BOWLS -2 2.95 15005 United Kingdom 2011-01-09 C541415 21527 RED RETROSPOT TRADITIONAL TEAPOT -1 7.95 14292 United Kingdom 2011-01-17 C537398 22430 ENAMEL WATERING CAN CREAM -1 4.95 15426 United Kingdom 2010-12-06 C537402 22196 SMALL HEART MEASURING SPOONS -12 0.85 15502 United Kingdom 2010-12-06 C5381

Notice that the quantities of product purchased are negative values

In [0]:
# number of orders that were cancelled

n_cancelled = cancelled.select('InvoiceNo').distinct().count()

print("Number of cancelled orders: ", n_cancelled)

Number of cancelled orders:  3654


#### StockCodes with alphabets

A look at the data shows there are stock codes that contain alphabets in them while some do not. We narrow down to this to understand more about this phenomenon.

From the dataset we note the meaning of these stock codes as below:

- POST -> Postage
- D -> Discount
- C2 -> Carriage
- M -> Manual
- BANK CHARGES -> Bank Charges
- PADS -> Pads to match all the cushions sold
- DOT -> Dotcom postage

In [0]:
# get stock codes with alphabets
alphaStockCodes = df1.filter(df1.StockCode.rlike('^[a-zA-Z]'))
display(alphaStockCodes)

InvoiceNo StockCode Description Quantity UnitPrice CustomerID Country dateFormatted 537804 M Manual 12 0.19 12748 United Kingdom 2010-12-08 538003 POST POSTAGE 8 18.0 12429 Denmark 2010-12-09 539840 POST POSTAGE 3 15.0 12383 Belgium 2010-12-22 540073 M Manual 6 1.65 16814 United Kingdom 2011-01-04 537915 POST POSTAGE 1 28.0 12797 Portugal 2010-12-09 540410 POST POSTAGE 3 18.0 12530 Germany 2011-01-07 540278 M Manual 1 1.65 15719 United Kingdom 2011-01-06 540351 POST POSTAGE 2 18.0 12735 France 2011-01-06 541405 POST POSTAGE 5 18.0 12683 France 2011-01-17 537201 POST POSTAGE 6 18.0 12472 Germany 2010-12-05 539421 C2 CARRIAGE 1 50.0 14016 EIRE 2010-12-17 536569 M Manual 1 18.95 16274 United Kingdom 2010-12-01 537463 POST POSTAGE 4 18.0 12681 France 2010-12-07 540688 POST POSTAGE 5 18.0 12736 France 2011-01-11 537208 M Manual 4 0.85 15889 United Kingdom 2010-12-05 C538897 D Discount -1 5.76 16422 United Kingdom 2010-12-15 539317 M Manual 2 3.0 17392 United Kingdom 2010-12-16 C538897 D Discount -1 42.5 16422 United Kingdom 2010-12-15 539636 POST POSTAGE 6 18.0 12481 Germany 2010-12-20 541857 POST POSTAGE 6 18.0 12727 France 2011-01-23 537892 POST POSTAGE 1 18.0 12481 Germany 2010-12-09 538196 POST POSTAGE 8 18.0 12731 France 2010-12-10 538175 POST POSTAGE 21 18.0 12471 Germany 2010-12-10 537246 POST POSTAGE 1 28.0 12808 Portugal 2010-12-06 537594 POST POSTAGE 5 18.0 12647 Germany 2010-12-07 542128 POST POSTAGE 2 18.0 12609 Germany 2011-01-25 539405 POST POSTAGE 6 18.0 12621 Germany 2010-12-17 541430 POST POSTAGE 18 18.0 12356 Portugal 2011-01-18 541980 POST POSTAGE 1 28.0 12421 Spain 2011-01-24 537198 POST POSTAGE 1 18.0 12647 Germany 2010-12-05 539946 POST POSTAGE 1 18.0 13817 Germany 2010-12-23 541265 M Manual 2 1.25 17609 United Kingdom 2011-01-16 541122 POST POSTAGE 2 18.0 12708 Germany 2011-01-14 542080 POST POSTAGE 5 18.0 13815 Germany 2011-01-25 538597 M Manual 1 1.25 15555 United Kingdom 2010-12-13 C537857 D Discount -1 267.12 17340 United Kingdom 2010-12-08 540480 POST POSTAGE 4 18.0 12480 Germany 2011-01-07 540365 POST POSTAGE 3 18.0 12413 France 2011-01-06 540789 POST POSTAGE 2 18.0 12643 France 2011-01-11 541491 POST POSTAGE 3 28.0 12510 Spain 2011-01-18 537963 C2 CARRIAGE 1 50.0 13369 United Kingdom 2010-12-09 C539712 POST POSTAGE -1 5.75 15602 United Kingdom 2010-12-21 542398 M Manual 1 4.0 13174 United Kingdom 2011-01-27 539353 POST POSTAGE 3 28.0 12782 Portugal 2010-12-17 541269 POST POSTAGE 8 18.0 12626 Germany 2011-01-17 542229 POST POSTAGE 1 18.0 12474 Germany 2011-01-26 539491 POST POSTAGE 1 15.0 14646 Netherlands 2010-12-20 537140 M Manual 1 0.42 12748 United Kingdom 2010-12-05 541569 POST POSTAGE 3 40.0 13520 Switzerland 2011-01-19 542375 POST POSTAGE 2 18.0 12626 Germany 2011-01-27 540455 POST POSTAGE 3 18.0 12583 France 2011-01-07 536370 POST POSTAGE 3 18.0 12583 France 2010-12-01 536861 POST POSTAGE 3 18.0 12427 Germany 2010-12-03 541485 M Manual 12 0.2 18118 United Kingdom 2011-01-18 539056 M Manual 1 1.45 16725 United Kingdom 2010-12-15 540839 M Manual 1 12.75 16791 United Kingdom 2011-01-11 542369 POST POSTAGE 1 18.0 12645 Germany 2011-01-27 540801 POST POSTAGE 6 28.0 12540 Spain 2011-01-11 537197 POST POSTAGE 3 18.0 12647 Germany 2010-12-05 539676 POST POSTAGE 3 15.0 12423 Belgium 2010-12-21 C536379 D Discount -1 27.5 14527 United Kingdom 2010-12-01 537212 POST POSTAGE 4 18.0 12720 Germany 2010-12-05 539607 POST POSTAGE 5 18.0 12681 France 2010-12-20 539984 C2 CARRIAGE 1 50.0 14911 EIRE 2010-12-23 539447 POST POSTAGE 3 18.0 12395 Belgium 2010-12-17 540546 POST POSTAGE 3 28.0 12766 Portugal 2011-01-09 540519 POST POSTAGE 2 28.0 12793 Portugal 2011-01-09 540769 POST POSTAGE 1 18.0 12601 Germany 2011-01-11 539327 POST POSTAGE 1 18.0 12600 Germany 2010-12-17 540040 POST POSTAGE 5 40.0 12483 Sweden 2011-01-04 536983 POST POSTAGE 1 18.0 12712 Germany 2010-12-03 540824 POST POSTAGE 2 18.0 12728 France 2011-01-11 541675 POST POSTAGE 2 15.0 12402 Belgium 2011-01-20 536403 POST POSTAGE 1 15.0 12791 Netherlands 2

Output can only be rendered in Databricks

#### Basket prices

In [0]:
# we create a new df with a new column TotalPrice derived from UnitPrice and Quantity
df1 = df1.withColumn("TotalPrice", df1.UnitPrice*df1.Quantity)

mvBasketDF = df1.groupBy('CustomerID','InvoiceNo').agg(sum('TotalPrice').alias('BasketPrice'))
mvBasketDF.show()

+----------+---------+------------------+
|CustomerID|InvoiceNo|       BasketPrice|
+----------+---------+------------------+
|     14723|   539321|208.46000000000004|
|     12601|   540769|            144.26|
|     13328|   542397|           1308.48|
|     16931|   539414|            212.61|
|     17609|   541265|             423.6|
|     15426|   537196|            382.57|
|     17850|   536690|            331.26|
|     15955|   536635| 572.3800000000001|
|     13018|   548920|493.45000000000005|
|     14796|   547730|349.51000000000005|
|     17431|   542613|403.29999999999995|
|     17811|   543801|             210.7|
|     13662|   547803|227.70000000000005|
|     13394|   543054|313.76000000000005|
|     14688|   548742|375.15999999999997|
|     13819|   543658|425.05000000000007|
|     16722|   544914|            314.36|
|     17449|   545338|338.34000000000003|
|     14163|   550180|            411.87|
|     15696|  C552707|             -9.95|
+----------+---------+------------

# Product Categories

#### TF-IDF
- [Term frequency-inverse document frequency (TF-IDF)](https://en.wikipedia.org/wiki/Tf%E2%80%93idf) is a feature vectorization method widely used in text mining to reflect the importance of a term to a document in the corpus. You can find more information about its implementation on PySpark [here](https://spark.apache.org/docs/latest/ml-features.html#tf-idf).

In [0]:
from pyspark.ml.feature import HashingTF, IDF, Tokenizer

tokenizer = Tokenizer(inputCol="Description", outputCol="descriptionTokens")
wordsData = tokenizer.transform(df1)

# tf transformation
hashingTF = HashingTF(inputCol="descriptionTokens", outputCol="rawFeatures", numFeatures=20)

featurizedData = hashingTF.transform(wordsData)

# alternatively, CountVectorizer can also be used to get term frequency vectors

# idf transformation
idf = IDF(inputCol="rawFeatures", outputCol="features")
idfModel = idf.fit(featurizedData)
rescaledData = idfModel.transform(featurizedData)

#### Creating clusters of products

In [0]:
# Import the required libraries
from pyspark.ml.clustering import KMeans
from pyspark.ml.feature import VectorAssembler
from pyspark.ml import Pipeline

# Create an object for the Logistic Regression model
kmeans_model = KMeans(k=4)

# fit
fit_model = kmeans_model.fit(rescaledData.select('features'))

# wsse = fit_model.computeCost(final_data) for spark 2.7
wssse = fit_model.summary.trainingCost # for spark 3.0
print("The within set sum of squared error of the mode is {}".format(wssse))

# Store the results in a dataframe

results = fit_model.transform(rescaledData.select('features'))

The within set sum of squared error of the mode is 3531130.7557436135


#### Finding optimal number of clusters

- **NOTE:** 
The part below DOES NOT scale. We could not find a suitable way of going about this step.

In [0]:
#silhoutte analysis to find the optimal number of clusters. 
silhouette_score=[]

from pyspark.ml.evaluation import ClusteringEvaluator

evaluator = ClusteringEvaluator(predictionCol='prediction', featuresCol='features',metricName='silhouette', distanceMeasure='squaredEuclidean')

for i in range(2,10):  
    KMeans_model = KMeans(k=i)
    KMeans_fit=KMeans_model.fit(rescaledData)
    output=KMeans_fit.transform(rescaledData)
    score=evaluator.evaluate(output) 
    silhouette_score.append(score)    
    print("Silhouette Score:", score)

Silhouette Score: 0.0976008504262662
Silhouette Score: 0.14821852091401963
Silhouette Score: 0.15049483139255138
Silhouette Score: 0.1603341734663381
Silhouette Score: 0.16864840283191324
Silhouette Score: 0.19020784681942218
Silhouette Score: 0.16420181156094876
Silhouette Score: 0.18791031822847862


- The silhoutte score is maximized when k = 8 but the difference between average silhoutte when k = 8 and k = 4 is very small. And, since the difference between the silhouette score is very small, Clustering with k = 8 may not group the points together that has similar characterstics than does k = 4. So, we are using clusters with 4 k's.

In [0]:
display(results.groupby('prediction').count().sort('prediction'))

prediction count 0 52421 1 45430 2 238623 3 65129

Output can only be rendered in Databricks

# Customer Categories

#### New feature: Product Category

In [0]:
# we add the product category allocations to the dataframe
from pyspark.sql.types import StructType, StructField, LongType

def with_column_index(df): 
    new_schema = StructType(df.schema.fields + [StructField("ColumnIndex", LongType(), False),])
    return df.rdd.zipWithIndex().map(lambda row: row[0] + (row[1],)).toDF(schema=new_schema)

df1_ci = with_column_index(df1)
df2_ci = with_column_index(results)

join_on_index = df1_ci.join(df2_ci, df1_ci.ColumnIndex == df2_ci.ColumnIndex, 'inner').drop("ColumnIndex","features","StockCode","Quantity","UnitPrice","Country","dateFormatted")

join_on_index.sort('InvoiceNo').show()

+---------+--------------------+----------+------------------+----------+
|InvoiceNo|         Description|CustomerID|        TotalPrice|prediction|
+---------+--------------------+----------+------------------+----------+
|   536365| WHITE METAL LANTERN|     17850|             20.34|         2|
|   536365|CREAM CUPID HEART...|     17850|              22.0|         2|
|   536365|RED WOOLLY HOTTIE...|     17850|             20.34|         3|
|   536365|KNITTED UNION FLA...|     17850|             20.34|         3|
|   536365|SET 7 BABUSHKA NE...|     17850|              15.3|         1|
|   536365|WHITE HANGING HEA...|     17850|15.299999999999999|         0|
|   536365|GLASS STAR FROSTE...|     17850|              25.5|         3|
|   536366|HAND WARMER RED P...|     17850|11.100000000000001|         3|
|   536366|HAND WARMER UNION...|     17850|11.100000000000001|         2|
|   536367|IVORY KNITTED MUG...|     13047| 9.899999999999999|         3|
|   536367|ASSORTED COLOUR B...|     1

#### Monetary value per category

In [0]:
# create columns from product category and fill null values with 0
pivotDF = join_on_index.groupBy("CustomerID","InvoiceNo","Description").pivot("prediction").sum("TotalPrice").na.fill(value=0)
display(pivotDF)

CustomerID InvoiceNo Description 0 1 2 3 16697 549910 RECYCLED ACAPULCO MAT TURQUOISE 8.25 0.0 0.0 0.0 16320 552695 HANGING JAM JAR T-LIGHT HOLDER 20.4 0.0 0.0 0.0 17193 556972 SET OF 4 KNICK KNACK TINS LONDON 0.0 0.0 49.800000000000004 0.0 16550 537391 PINK POLKADOT CUP 0.0 0.0 0.0 5.95 12415 543989 CHILDS BREAKFAST SET SPACEBOY 0.0 0.0 204.0 0.0 14911 545657 JAM MAKING SET WITH JARS 0.0 0.0 22.5 0.0 15750 557745 UNION JACK FLAG PASSPORT COVER 0.0 0.0 0.0 6.300000000000001 17511 542789 WRAP ALPHABET DESIGN 0.0 0.0 10.5 0.0 18192 568792 DOORMAT SPOTTY HOME SWEET HOME 0.0 16.5 0.0 0.0 16628 561391 WOODLAND CHARLOTTE BAG 0.0 0.0 8.5 0.0 12701 567578 PACK OF 12 PAISLEY PARK TISSUES 0.0 0.0 9.36 0.0 17220 548308 QUEENS GUARD COFFEE MUG 15.299999999999999 0.0 0.0 0.0 16327 537137 ZINC METAL HEART DECORATION 0.0 0.0 6.25 0.0 17856 545505 ANTIQUE SILVER TEA GLASS ENGRAVED 0.0 0.0 0.0 305.28000000000003 13869 562688 BLUE POLKADOT WRAP 0.0 0.0 10.5 0.0 13089 580673 LUNCH BAG BLACK SKULL. 0.0 0.0 0.0 33.0 12427 570452 SPACEBOY CHILDRENS EGG CUP 0.0 0.0 15.0 0.0 12395 575151 RECIPE BOX RETROSPOT 0.0 0.0 17.700000000000003 0.0 14298 568675 JUMBO STORAGE BAG SKULLS 0.0 0.0 53.7 0.0 13230 545460 PAINT YOUR OWN CANVAS SET 0.0 0.0 3.3 0.0 15386 555649 MODERN FLORAL STATIONERY SET 0.0 15.0 0.0 0.0 13211 565664 WRAP VINTAGE DOILY 0.0 0.0 0.0 10.5 16161 539034 POPPY'S PLAYHOUSE BEDROOM 8.4 0.0 0.0 0.0 17357 576611 ASSORTED TUTTI FRUTTI NOTEBOOK 0.0 4.2 0.0 0.0 13018 566006 CHRISTMAS CRAFT TREE TOP ANGEL 0.0 12.600000000000001 0.0 0.0 15838 569815 WOODEN PICTURE FRAME WHITE FINISH 151.20000000000002 0.0 0.0 0.0 15044 577519 6 ROCKET BALLOONS 0.0 0.0 6.5 0.0 15819 561896 JUMBO BAG ALPHABET 0.0 0.0 20.8 0.0 15076 557319 HAND OVER THE CHOCOLATE SIGN 0.0 0.0 0.0 25.200000000000003 15984 563924 WASHROOM METAL SIGN 0.0 0.0 1.45 0.0 16767 565790 LUNCH BAG BLACK SKULL. 0.0 0.0 0.0 16.5 16554 C580966 HEART WREATH DECORATION WITH BELL 0.0 0.0 -15.0 0.0 12757 569870 PLAYING CARDS I LOVE LONDON 0.0 0.0 2.5 0.0 15214 573770 VINTAGE CHRISTMAS TABLECLOTH 0.0 25.0 0.0 0.0 15506 561034 METAL SIGN TAKE IT OR LEAVE IT 0.0 0.0 5.9 0.0 14562 566073 ENGLISH ROSE HOT WATER BOTTLE 0.0 0.0 8.5 0.0 14606 569457 ORANGE SCENTED SET/9 T-LIGHTS 0.0 0.85 0.0 0.0 16385 537792 HERB MARKER MINT 0.0 0.0 0.65 0.0 14481 558295 12 PENCILS SMALL TUBE SKULL 0.0 0.0 15.600000000000001 0.0 16056 572726 DOG BOWL CHASING BALL DESIGN 0.0 0.0 0.0 22.5 15146 544571 DECORATION PINK CHICK MAGIC GARDEN 0.0 13.6 0.0 0.0 14298 544296 RETRO PILL BOX , REVOLUTIONARY 0.0 18.0 0.0 0.0 13176 572528 PAPER POCKET TRAVELING FAN 0.0 0.0 3.3600000000000003 0.0 14194 546988 ALARM CLOCK BAKELIKE GREEN 0.0 0.0 15.0 0.0 17070 563585 SET OF 36 DINOSAUR PAPER DOILIES 1.45 0.0 0.0 0.0 16773 571062 FEATHER PEN,HOT PINK 0.0 0.0 23.400000000000002 0.0 18276 572990 ICE CREAM SUNDAE LIP GLOSS 0.0 30.0 0.0 0.0 17426 557217 ROBOT BIRTHDAY CARD 0.0 0.0 5.04 0.0 14606 541859 WOOD S/3 CABINET ANT WHITE FINISH 0.0 7.95 0.0 0.0 13199 580031 WALL ART ONLY ONE PERSON 0.0 0.0 74.7 0.0 13694 564323 JUMBO BAG WOODLAND ANIMALS 0.0 0.0 175.0 0.0 14096 578270 BUNDLE OF 3 RETRO NOTE BOOKS 0.0 3.29 0.0 0.0 12836 551855 ASSORTED COLOUR BIRD ORNAMENT 0.0 0.0 40.56 0.0 14096 575607 DOORMAT FAIRY CAKE 15.79 0.0 0.0 0.0 14390 550130 PHOTO CUBE 0.0 0.0 9.899999999999999 0.0 12444 577314 FRENCH ENAMEL CANDLEHOLDER 0.0 0.0 25.200000000000003 0.0 16871 553067 DOORMAT UNION FLAG 0.0 0.0 0.0 47.7 14907 560120 16 PIECE CUTLERY SET PANTRY DESIGN 0.0 0.0 95.69999999999999 0.0 14401 580650 LARGE RED BABUSHKA NOTEBOOK 0.0 0.0 0.0 4.68 14527 549287 DOORMAT I LOVE LONDON 0.0 0.0 7.95 0.0 16243 571269 SPOTTY BUNTING 0.0 0.0 4.95 0.0 17075 542226 PEARL CRYSTAL PUMPKIN T-LIGHT HLDR 0.0 0.0 9.899999999999999 0.0 17614 571105 GARDENIA 1 WICK MORRIS BOXED CANDLE 0.0 0.0 3.75 0.0 15311 541575 RED RETROSPOT MUG 0.0 0.0 0.0 2.95 14396 569702 EMBROIDERED RIBBON REEL SALLY 0.0 0.0 12.48 0.0 14235 568192 PLAYING CARDS KEEP CALM & CARRY ON 0.0 0.0 30.0 0.0 17802 556

In [0]:
mvBasketDF.sort('CustomerID').show()

+----------+---------+------------------+
|CustomerID|InvoiceNo|       BasketPrice|
+----------+---------+------------------+
|     12346|  C541433|          -77183.6|
|     12346|   541431|           77183.6|
|     12347|   556201|382.52000000000004|
|     12347|   562032| 584.9100000000001|
|     12347|   573511|           1294.32|
|     12347|   537626| 711.7900000000001|
|     12347|   549222|            636.25|
|     12347|   581180|            224.82|
|     12347|   542237|            475.39|
|     12348|   541998|227.43999999999997|
|     12348|   539318|             892.8|
|     12348|   568172|             310.0|
|     12348|   548955|             367.0|
|     12349|   577609|           1757.55|
|     12350|   543037|334.40000000000003|
|     12352|   544156|             296.5|
|     12352|   546869|            120.33|
|     12352|   547390|160.32999999999998|
|     12352|   545332|             840.3|
|     12352|   567505|366.25000000000006|
+----------+---------+------------

#### Value of each category per basket

In [0]:
# value distribution by product category
basketValPerCat = pivotDF.groupBy('CustomerID','InvoiceNo').agg(sum('0').alias('cat0'),sum('1').alias('cat1'),sum('2').alias('cat2'),sum('3').alias('cat3'))

# total basket value as a new column
basketValPerCat = basketValPerCat.withColumn("BasketValue", basketValPerCat.cat0 + basketValPerCat.cat1 + basketValPerCat.cat2 + basketValPerCat.cat3)

# drop rows whose total basket value is negative (ie cancellations)
basketValPerCat = basketValPerCat.where(basketValPerCat.BasketValue > 0)
basketValPerCat.show()

+----------+---------+------------------+------------------+------------------+------------------+------------------+
|CustomerID|InvoiceNo|              cat0|              cat1|              cat2|              cat3|       BasketValue|
+----------+---------+------------------+------------------+------------------+------------------+------------------+
|     16056|   572726|              88.4|60.150000000000006|435.30999999999995|             66.89| 650.7499999999999|
|     13328|   542397|              30.0| 79.19999999999999|             889.8|            309.48|           1308.48|
|     13018|   548920|             160.1|              81.6|             200.8|             50.95|            493.45|
|     12700|   569568|107.30000000000001|            270.18| 907.7099999999998|             191.0|1476.1899999999998|
|     14796|   547730|             77.05|             75.94|            126.02|              70.5|            349.51|
|     16931|   560244|28.269999999999996|             22

#### Customer data

In [0]:
from pyspark.sql.functions import count, min, sum, max, avg

customersDF = basketValPerCat.groupBy('CustomerID').agg(count('InvoiceNo').alias('no_purchases'),min('BasketValue').alias('minVal'),max('BasketValue').alias('maxVal'),\
                                                        avg('BasketValue').alias('meanVal'),sum('BasketValue').alias('totalVal'),sum('cat0').alias('cat0'),sum('cat1').alias('cat1'),sum('cat2').alias('cat2'),sum('cat3').alias('cat3'))

In [0]:
# distribution of total value of purchases per customer

display(customersDF)

CustomerID no_purchases minVal maxVal meanVal totalVal cat0 cat1 cat2 cat3 17389 34 52.5 7427.32 936.2847058823529 31833.68 440.94 243.89999999999998 25600.619999999995 5548.219999999999 14450 3 108.42000000000002 213.8 161.08333333333334 483.25 17.58 84.0 266.67 115.0 15727 7 374.82000000000005 1516.95 737.0085714285714 5159.0599999999995 767.6300000000001 193.20999999999998 3277.66 920.5600000000001 15790 1 218.75 218.75 218.75 218.75 14.22 36.5 111.09 56.94 13285 4 506.38 796.8299999999999 677.28 2709.12 256.41999999999996 437.9 1602.56 412.24 14570 2 50.150000000000006 167.91 109.03 218.06 38.8 13.95 143.91 21.4 16574 1 451.44000000000005 451.44000000000005 451.44000000000005 451.44000000000005 43.32 15.299999999999999 303.72 89.1 13832 1 52.199999999999996 52.199999999999996 52.199999999999996 52.199999999999996 32.4 0.0 19.799999999999997 0.0 13623 5 88.6 188.57 145.54799999999997 727.7399999999999 103.01 158.73 425.09999999999997 40.900000000000006 15619 1 336.40000000000003 336.40000000000003 336.40000000000003 336.40000000000003 0.0 91.80000000000001 99.60000000000001 145.0 15967 2 161.9 235.78 198.84 397.68 48.81 25.11 283.98 39.78 14465 3 186.46 597.3 346.1466666666667 1038.44 141.65 38.1 750.94 107.75 18161 5 224.94 462.00000000000006 323.73799999999994 1618.6899999999998 189.86 12.48 1332.35 84.0 12471 30 29.700000000000003 2238.51 659.6216666666667 19788.65 1639.0600000000002 2351.62 14108.7 1689.27 12626 9 178.1 1377.55 735.6088888888888 6620.48 639.0899999999999 940.8800000000001 4107.57 932.94 13261 2 99.0 845.1 472.05 944.1 0.0 0.0 944.1 0.0 14148 1 157.2 157.2 157.2 157.2 0.0 0.0 157.2 0.0 14423 1 265.2 265.2 265.2 265.2 265.2 0.0 0.0 0.0 14536 1 170.95 170.95 170.95 170.95 10.85 0.0 137.05 23.049999999999997 18218 2 289.22 352.7 320.96000000000004 641.9200000000001 43.2 0.0 368.22 230.5 18282 2 77.84 100.21000000000001 89.025 178.05 22.95 0.0 142.35000000000002 12.75 16500 6 30.0 165.0 66.81 400.86 56.760000000000005 165.0 149.10000000000002 30.0 15100 3 175.2 350.4 292.0 876.0 876.0 0.0 0.0 0.0 15738 16 16.6 391.8 300.77625 4812.42 563.4000000000001 219.27999999999997 2561.84 1467.8999999999999 17837 8 119.79999999999998 462.33 192.65375 1541.23 313.7 60.53 964.92 202.07999999999998 13601 2 320.12 373.2 346.65999999999997 693.3199999999999 70.67999999999999 67.07 479.14 76.43 14289 6 78.48 322.24 151.8766666666667 911.2600000000001 47.0 62.0 661.56 140.7 16791 7 13.52 479.36 243.6828571428571 1705.7799999999997 222.48999999999998 233.00000000000003 1075.54 174.75 15655 1 560.67 560.67 560.67 560.67 194.7 17.9 278.92 69.15 15296 6 13.899999999999999 1199.33 612.1433333333333 3672.8599999999997 573.6 366.83 2250.31 482.12 17754 5 191.63 523.46 332.93 1664.65 205.42000000000002 68.7 805.02 585.51 18221 5 162.95 302.53999999999996 208.32200000000003 1041.6100000000001 267.02 46.5 564.19 163.89999999999998 12393 4 177.00000000000003 521.5 395.65 1582.6 23.4 71.1 1037.3000000000002 450.8 13875 5 340.65 384.19999999999993 357.358 1786.79 180.57999999999998 123.3 1210.91 272.0 16924 3 252.54000000000002 306.24 276.63 829.89 24.6 81.68 645.44 78.17 14420 1 387.52 387.52 387.52 387.52 59.64 10.08 268.4 49.39999999999999 17809 12 34.8 1835.6000000000001 450.99250000000006 5411.910000000001 867.34 965.25 2411.0 1168.3200000000002 14997 3 112.52 278.4 182.06666666666663 546.1999999999999 39.78 104.39999999999999 307.92 94.10000000000001 14377 2 302.45000000000005 314.6 308.52500000000003 617.0500000000001 219.2 31.8 231.05 135.0 16224 4 167.47 429.41999999999996 323.825 1295.3 154.04 210.83999999999997 776.97 153.45 15382 6 26.400000000000002 2582.78 992.6183333333335 5955.710000000001 665.64 653.54 3770.7799999999997 865.75 15604 1 216.39000000000004 216.39000000000004 216.39000000000004 216.39000000000004 28.24 0.0 112.67000000000002 75.48 17223 2 124.05 292.84 208.445 416.89 22.34 0.0 331.19999999999993 63.349999999999994 13898 1 155.93 155.93 155.93 155.93 41.52 15.0 44.91 54.5 17193 7 16.54 760.74 242.757142857142

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

Output can only be rendered in Databricks

The distribution of each individual numerical column is highly skewed probably because there exists absurdly high values related to **cancelled/erronous orders** which were not removed from the dataset. This **will** significantly affect the generizability of the model to be trained on the data.

#### Creating customer categories

- In this step we use unsupervised learning to create customer categories.
- It is the basis of the next step - classification

In [0]:
# Vector assembler is used to create a vector of input features
assembler2 = VectorAssembler(inputCols=['no_purchases','minVal','maxVal','meanVal','totalVal','cat0','cat1','cat2','cat3'],\
                            outputCol="features")

# Pipeline is used to pass the data through indexer and assembler simultaneously. Also, it helps to pre-rocess the test data
# in the same way as that of the train data.
pipe2 = Pipeline(stages=[assembler2])

final_data2=pipe2.fit(customersDF).transform(customersDF)

# Create an object for the Logistic Regression model

kmeans_model2 = KMeans(k=4)

# fit
fit_model2 = kmeans_model2.fit(final_data2.select('features'))

# wsse = fit_model.computeCost(final_data) for spark 2.7
wssse2 = fit_model2.summary.trainingCost # for spark 3.0
print("The within set sum of squared error of the mode is {}".format(wssse2))

# Store the results in a dataframe

results2 = fit_model2.transform(final_data2.select('features'))

The within set sum of squared error of the mode is 98509463715.91588


In [0]:
display(results2.groupby('prediction').count().sort('prediction'))

prediction count 0 4212 1 4 2 104 3 18

Output can only be rendered in Databricks

In [0]:
# we add the customer category allocations to the customers dataframe
from pyspark.sql.types import StructType, StructField, LongType

def with_column_index(df): 
    new_schema = StructType(df.schema.fields + [StructField("ColumnIndex", LongType(), False),])
    return df.rdd.zipWithIndex().map(lambda row: row[0] + (row[1],)).toDF(schema=new_schema)

# using the function defined earlier
df11_ci = with_column_index(customersDF)
df21_ci = with_column_index(results2)

join_on_index1 = df11_ci.join(df21_ci, df11_ci.ColumnIndex == df21_ci.ColumnIndex, 'inner').drop("ColumnIndex","features","no_purchases","minVal","maxVal","totalVal")

#### Data visualization

In [0]:
display(join_on_index1)

CustomerID meanVal cat0 cat1 cat2 cat3 prediction 17389 936.2847058823529 440.94 243.89999999999998 25600.619999999995 5548.219999999999 2 14450 161.08333333333334 17.58 84.0 266.67 115.0 0 15727 737.0085714285714 767.6300000000001 193.20999999999998 3277.66 920.5600000000001 0 15790 218.75 14.22 36.5 111.09 56.94 0 13285 677.28 256.41999999999996 437.9 1602.56 412.24 0 14570 109.03 38.8 13.95 143.91 21.4 0 16574 451.44000000000005 43.32 15.299999999999999 303.72 89.1 0 13832 52.199999999999996 32.4 0.0 19.799999999999997 0.0 0 13623 145.54799999999997 103.01 158.73 425.09999999999997 40.900000000000006 0 15619 336.40000000000003 0.0 91.80000000000001 99.60000000000001 145.0 0 15967 198.84 48.81 25.11 283.98 39.78 0 14465 346.1466666666667 141.65 38.1 750.94 107.75 0 18161 323.73799999999994 189.86 12.48 1332.35 84.0 0 12471 659.6216666666667 1639.0600000000002 2351.62 14108.7 1689.27 2 12626 735.6088888888888 639.0899999999999 940.8800000000001 4107.57 932.94 0 13261 472.05 0.0 0.0 944.1 0.0 0 14148 157.2 0.0 0.0 157.2 0.0 0 14423 265.2 265.2 0.0 0.0 0.0 0 14536 170.95 10.85 0.0 137.05 23.049999999999997 0 18218 320.96000000000004 43.2 0.0 368.22 230.5 0 18282 89.025 22.95 0.0 142.35000000000002 12.75 0 16500 66.81 56.760000000000005 165.0 149.10000000000002 30.0 0 15100 292.0 876.0 0.0 0.0 0.0 0 15738 300.77625 563.4000000000001 219.27999999999997 2561.84 1467.8999999999999 0 17837 192.65375 313.7 60.53 964.92 202.07999999999998 0 13601 346.65999999999997 70.67999999999999 67.07 479.14 76.43 0 14289 151.8766666666667 47.0 62.0 661.56 140.7 0 16791 243.6828571428571 222.48999999999998 233.00000000000003 1075.54 174.75 0 15655 560.67 194.7 17.9 278.92 69.15 0 15296 612.1433333333333 573.6 366.83 2250.31 482.12 0 17754 332.93 205.42000000000002 68.7 805.02 585.51 0 18221 208.32200000000003 267.02 46.5 564.19 163.89999999999998 0 12393 395.65 23.4 71.1 1037.3000000000002 450.8 0 13875 357.358 180.57999999999998 123.3 1210.91 272.0 0 16924 276.63 24.6 81.68 645.44 78.17 0 14420 387.52 59.64 10.08 268.4 49.39999999999999 0 17809 450.99250000000006 867.34 965.25 2411.0 1168.3200000000002 0 14997 182.06666666666663 39.78 104.39999999999999 307.92 94.10000000000001 0 14377 308.52500000000003 219.2 31.8 231.05 135.0 0 16224 323.825 154.04 210.83999999999997 776.97 153.45 0 15382 992.6183333333335 665.64 653.54 3770.7799999999997 865.75 0 15604 216.39000000000004 28.24 0.0 112.67000000000002 75.48 0 17223 208.445 22.34 0.0 331.19999999999993 63.349999999999994 0 13898 155.93 41.52 15.0 44.91 54.5 0 17193 242.7571428571429 136.48 150.66 1412.16 0.0 0 14713 242.0181818181819 411.1 270.05 1639.13 341.91999999999996 0 15292 178.39 77.69999999999999 11.8 82.99 5.9 0 17190 222.37 37.35000000000001 20.310000000000002 372.29999999999995 14.780000000000001 0 14944 491.6875 893.88 416.65999999999997 3933.7700000000004 655.94 0 16999 242.41000000000003 0.0 15.0 399.53000000000003 70.28999999999999 0 17950 160.73333333333332 50.800000000000004 36.09 305.58 89.73 0 16117 232.21 67.6 19.57 109.58000000000001 35.46 0 16706 177.11666666666665 89.04999999999998 47.550000000000004 266.63 128.12 0 15740 154.88 37.44 16.6 83.5 17.34 0 13225 923.4057142857145 1567.7800000000002 1284.66 2422.68 1188.72 0 14779 128.71666666666667 42.150000000000006 43.05 229.59999999999997 71.35 0 13468 154.18837837837836 1792.2599999999995 608.99 2705.33 598.39 0 17346 174.85866666666666 314.39 354.34000000000003 1379.04 575.11 0 16549 403.49 468.09999999999997 558.91 2371.03 636.86 0 15026 156.33333333333334 33.5 44.18 374.32 17.0 0 15146 663.58 143.04 274.86 752.5 156.76 0 16295 332.1375 292.0 153.26 633.6 249.69 0 16985 546.412 1443.3 745.4499999999999 2317.2799999999997 958.0899999999999 0 17438 388.45000000000005 0.0 37.3 217.3 133.85000000000002 0 16526 290.36 0.0 31.220000000000002 212.14 47.0 0 16506 45.150000000000006 0.0 0.0 90.30000000000001 0.0 0 13500 388.92333333333335 139.26999999999998 166.91 721.37 139.22000000000003 0 12694 434.83500000000004 23.14 106.38 

# Classification of Customers
- We use only a few of the features from earlier steps in this next step to classify customers.

#### Train-test split

In [0]:
data = join_on_index1.withColumnRenamed("prediction", "customerGroup")

train_data,test_data=data.randomSplit([0.7,0.3])

#### Classifier: Decision Tree

In [0]:
# Import the required libraries
from pyspark.ml.classification import DecisionTreeClassifier
from pyspark.ml.feature import VectorAssembler
from pyspark.ml import Pipeline

In [0]:
# Vector assembler is used to create a vector of input features
assembler = VectorAssembler(inputCols=['CustomerID', 'meanVal', 'cat0', 'cat1', 'cat2', 'cat3'],
                            outputCol="features")

# Create an object for the decision tree classifier
dt_model = DecisionTreeClassifier(labelCol='customerGroup',maxBins=5000)

pipe = Pipeline(stages=[assembler,dt_model])

# fit model
fit_model=pipe.fit(train_data)

# Store the results in a dataframe
results = fit_model.transform(test_data)

# evaluate model
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

ACC_evaluator = MulticlassClassificationEvaluator(labelCol="customerGroup", predictionCol="prediction", metricName="accuracy")
accuracy = ACC_evaluator.evaluate(results)

print("The accuracy of the decision tree classifier is {}".format(accuracy))

The accuracy of the decision tree classifier is 0.9947328818660647


#### Classifier: Random Forest

In [0]:
from pyspark.ml.classification import RandomForestClassifier

# RandomForest model.
rf = RandomForestClassifier(labelCol="customerGroup", numTrees=10)

# Chain indexers and forest in a Pipeline
pipeRF = Pipeline(stages=[assembler, rf])

# fit model
fit_modelRF = pipeRF.fit(train_data)

# Store the results in a dataframe
resultsRF = fit_modelRF.transform(test_data)

# evaluate model
accuracyRF = ACC_evaluator.evaluate(resultsRF)

print("The accuracy of the random forest classifier is {}".format(accuracyRF))

The accuracy of the random forest classifier is 0.9917231000752446
